### ISSUE TITLE
Poorly conditioned matrices are still leading to numerical issues

### ISSUE THREAD

**Issue by samreid** [phetsims/circuit-construction-kit-common#744]
From https://github.com/phetsims/circuit-construction-kit-common/issues/693#issuecomment-927009451 and below.

> @arouinfar and I were able to reproduce the problem.  It was very sensitive to a specific position/resistance of the wires.  It doesn't seem to be connected to the battery "shorting out" mode.  I'll look into it further, perhaps commenting out some of the new graph postprocessing steps to see if they are correlated.  But this is not reliable to reproduce every time, so may be difficult to know when there is a problem vs not.

![Kapture 2021-09-24 at 22 45 15](https://user-images.githubusercontent.com/679486/134758458-250f5f2d-c4d8-4957-bf91-935f48c5c579.gif)

I spent 5-10 minutes trying to reproduce the problem this morning.  Then on a design call with @arouinfar we were able to reproduce the problem in a few minutes.  But when swapping out a resistor for a fresh resistor, we couldn't reproduce it any more.  Then I spent another 5-10 minutes and could not reproduce the problem.

Next hypothesis to test is: running with ?dev check the resistances of the wires, because the problem appears to be triggered when changing the wire lengths.

I tested with ?dev and got a buggy case and found that all wire resistances look like good values.

I commented out the 2 graph postprocessing steps, and it did not affect the result.

I also confirmed for the buggy circuit, all elements are participants, and nothing is getting thrown out as a dangling branch.

**Comment by samreid** [comment]
All indications suggest this is another manifestation of the floating point precision problem we saw for poorly conditioned matrices in https://github.com/phetsims/circuit-construction-kit-common/issues/693.  I can simulate the problem by using this patch and turning down the precision.  Likewise, turning up the precision addresses the problem.

<details>

```diff

Index: js/CCKCQueryParameters.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/CCKCQueryParameters.js b/js/CCKCQueryParameters.js
--- a/js/CCKCQueryParameters.js	(revision 5a40dba98cdfaa067dd50908cdaf13ea0ad193f3)
+++ b/js/CCKCQueryParameters.js	(date 1632859658666)
@@ -97,7 +97,7 @@
   // Precision for the numerical matrix solution for the linear algebra step
   precision: {
     type: 'number',
-    defaultValue: 32
+    defaultValue: 64
   },
 
   // For debugging the current value and sense of FixedCircuitElement
Index: js/model/ModifiedNodalAnalysisCircuit.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/model/ModifiedNodalAnalysisCircuit.js b/js/model/ModifiedNodalAnalysisCircuit.js
--- a/js/model/ModifiedNodalAnalysisCircuit.js	(revision 5a40dba98cdfaa067dd50908cdaf13ea0ad193f3)
+++ b/js/model/ModifiedNodalAnalysisCircuit.js	(date 1632860939888)
@@ -21,10 +21,16 @@
 import circuitConstructionKitCommon from '../circuitConstructionKitCommon.js';
 import ModifiedNodalAnalysisSolution from './ModifiedNodalAnalysisSolution.js';
 
-const LUDecimal = Decimal.clone( {
-  // precision: 16 // default precision for {number}
-  precision: CCKCQueryParameters.precision
-} );
+let LUDecimal = null;
+
+window.dec = precision => {
+  LUDecimal = Decimal.clone( {
+    // precision: 16 // default precision for {number}
+    precision: precision
+  } );
+};
+
+window.dec( CCKCQueryParameters.precision );
 
 class ModifiedNodalAnalysisCircuit {
 
@@ -337,9 +343,10 @@
     try {
 
       CCKCUtils.incrementAccumulatedSteps();
+      console.log( CCKCUtils.getAccumulatedSteps() );
 
       // For the first several steps, use high precision
-      if ( CCKCUtils.getAccumulatedSteps() <= 3 ) {
+      if ( true ) {
 
         // Must be a plain array, not an optimized number container
         const entries = [];
@@ -368,6 +375,8 @@
       // debugger yet to understand the cause.  Catch it and provide a solution of zeroes of the correct dimension
       // See https://github.com/phetsims/circuit-construction-kit-dc/issues/113
       x = new Matrix( A.n, 1 );
+
+      console.log( 'deficient matrix rank' );
     }
 
     // The matrix should be square since it is an exact analytical solution, see https://github.com/phetsims/circuit-construction-kit-dc/issues/96
@@ -376,6 +385,9 @@
     if ( phet.log ) {
       console.log( getDebugInfo( this, A, z, equations, unknowns, x ) );
     }
+    else {
+      console.log( A.cond() );
+    }
 
     const voltageMap = {};
     for ( let i = 0; i < unknownVoltages.length; i++ ) {
Index: js/model/TimestepSubdivisions.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/model/TimestepSubdivisions.js b/js/model/TimestepSubdivisions.js
--- a/js/model/TimestepSubdivisions.js	(revision 5a40dba98cdfaa067dd50908cdaf13ea0ad193f3)
+++ b/js/model/TimestepSubdivisions.js	(date 1632859547168)
@@ -14,9 +14,11 @@
 
 // smallest possible time
 const MIN_DT = 1E-6;
+window.minDT = MIN_DT;
 
 //threshold for determining whether 2 states are similar enough; any error less than errorThreshold will be tolerated.
 const ERROR_THRESHOLD = 1E-5;
+window.errorThreshold = ERROR_THRESHOLD;
 
 /**
  * @typedef Steppable
@@ -40,6 +42,7 @@
     let attemptedDT = totalTime;
     while ( elapsedTime < totalTime ) {
 
+      console.log( 'elapsed time = ' + elapsedTime.toFixed( 5 ) + ', total time = ' + totalTime.toFixed( 5 ) );
       const result = this.search( state, steppable, attemptedDT, null );
       state = result.state;
       states.push( result );
@@ -70,8 +73,8 @@
   search( state, steppable, dt, halfStepState ) {
 
     // if dt is already too low, no need to do error checking
-    if ( dt <= MIN_DT ) {
-      return { dt: MIN_DT, state: steppable.update( state, MIN_DT ) };
+    if ( dt <= window.minDT ) {
+      return { dt: window.minDT, state: steppable.update( state, window.minDT ) };
     }
     else {
       const a = steppable.update( state, dt );
@@ -79,7 +82,7 @@
       const b2 = steppable.update( b1, dt / 2 );
       const distance = steppable.distance( a, b2 );
       assert && assert( !isNaN( distance ), 'distance should be numeric' );
-      const errorAcceptable = distance < ERROR_THRESHOLD;
+      const errorAcceptable = distance < window.errorThreshold;
       if ( errorAcceptable ) {
         return { dt: dt, state: b2 }; // Use the more precise estimate
       }
Index: js/model/ModifiedNodalAnalysisAdapter.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/model/ModifiedNodalAnalysisAdapter.js b/js/model/ModifiedNodalAnalysisAdapter.js
--- a/js/model/ModifiedNodalAnalysisAdapter.js	(revision 5a40dba98cdfaa067dd50908cdaf13ea0ad193f3)
+++ b/js/model/ModifiedNodalAnalysisAdapter.js	(date 1632858867474)
@@ -342,19 +342,23 @@
 
     // Start visiting from the solved vertices, since they have the ground truth.  Have to visit each to make sure
     // we traveled to all disconnected components
-    const allVertices = [ ...solvedVertices, ...unsolvedVertices ];
-    allVertices.forEach( vertex => {
-      if ( !visited.includes( vertex ) ) {
-        dfs( vertex, visitVoltage );
-      }
-    } );
-
-    // Depth first search across the circuit to ensure current conserved at each vertex
+    // const allVertices = [ ...solvedVertices, ...unsolvedVertices ];
+    // if ( window.voltages ) {
+    //   allVertices.forEach( vertex => {
+    //     if ( !visited.includes( vertex ) ) {
+    //       dfs( vertex, visitVoltage );
+    //     }
+    //   } );
+    // }
+    //
+    // // Depth first search across the circuit to ensure current conserved at each vertex
     // circuit.checkCurrentConservation( 'before' );
-    const locked = [ ...nonParticipants ];
-    const visitCurrent = vertex => circuit.conserveCurrent( vertex, locked );
-    visited.length = 0;
-    allVertices.forEach( vertex => dfs( vertex, visitCurrent ) );
+    // const locked = [ ...nonParticipants ];
+    // const visitCurrent = vertex => circuit.conserveCurrent( vertex, locked );
+    // visited.length = 0;
+    // if ( window.currents ) {
+    //   allVertices.forEach( vertex => dfs( vertex, visitCurrent ) );
+    // }
     // circuit.checkCurrentConservation( 'after' );
   }
 }

```

</details>

We cannot turn the precision too high or it wrecks the performance.  But if the precision is low, even the LUDecompositionDecimal will give a bad solution for these poorly conditioned matrices.  The condition number for this circuit is `107926455236.72458` and higher.  We also may be able to tolerate a higher precision if we are careful to only run it once on static circuits.  Likewise, for a linear circuit, we may speed up 66% by running timestep subdivisions once instead of 3 times for the base case.  But when adding AC components (AC, capacitor, inductor), we will have to run the solver every frame.

**Comment by samreid** [comment]
To understand why this problem seems new in HTML5, I outputted the matrix for a Java battery-resistor-wire circuit.  The result is:

```
 0.00 1.00 0.00 0.00 0.00
 0.00 10000.10 -0.10 0.00 -10000.00
 0.00 -0.10 5271.40 -5271.30 0.00
 1.00 0.00 -5271.30 5271.30 0.00
-1.00 -10000.00 0.00 0.00 10000.00
 0.00 0.00 0.00 -1.00 1.00
```

Note how this has a much better conditioning than the similar matrix for HTML5:

```
A.cond=1E10.4242 = 26559515060.5876  
A=
dim: 5x5
0 0 1 -1 0 
1 -9052643497.2 0 -1000000 0 
0 9052643497.3 -0.1 0 0 
0 -0.1 0.1 0 -1 
0 0 0 1000000 1 
```

I believe this is caused by the tiny resistances in the wire and battery in the HTML5 version.  Note I tried to compute the condition number of the Java version like so:

```js
const x = `
 0.00 1.00 0.00 0.00 0.00
 0.00 10000.10 -0.10 0.00 -10000.00
 0.00 -0.10 5271.40 -5271.30 0.00
 1.00 0.00 -5271.30 5271.30 0.00
-1.00 -10000.00 0.00 0.00 10000.00
 0.00 0.00 0.00 -1.00 1.00
`;

const y = x.trim().split( '\n' );
const m = new Matrix( 6, 6 );
for ( let i = 0; i < y.length; i++ ) {
  const row = y[ i ].trim().split( ' ' );
  for ( let k = 0; k < row.length; k++ ) {
    console.log( row[ k ] );
    m.set( i, k, parseFloat( row[ k ] ) );
  }
}
console.log( m.toString() );
console.log( m.cond() );
```
But it failed, perhaps because it is a non-square matrix.  That means it is being solved with the QR instead of LU decomposition.

**Comment by samreid** [comment]
Maybe we should experiment with QR for the reasons described in https://math.stackexchange.com/questions/2836110/why-which-advantages-we-use-different-matrix-factorization-algorithms

>  In general the normal LU decomp isn't used because of the stability.
> You would utilize the QR decomposition for numerical stability. It is the back bone of most techniques for orthogonalization.

Though generally, for square matrices square-optimal algorithms are typically used.

**Comment by Nancy-Salpepi** [comment]
**Test device**
MacBook Air (m1 chip)

**Operating System**
11.6

**Browser**
chrome

**Problem description**
I'm not positive this relates to this issue, but I was able to produce a current through a dollar bill by repositioning/changing the length of the wires. This is similar to issues commented on earlier by @stemilymill and myself in https://github.com/phetsims/circuit-construction-kit-common/issues/693.

I was easily able to reproduce this problem.

**Steps to reproduce**
Started at the dollar bill and moved in a counterclockwise direction to complete the circuit shown below. 

**Visuals**

![currentinbill](https://user-images.githubusercontent.com/87318828/135516810-98cab77c-41eb-45d6-927d-4d27ffd44c1f.gif)

Troubleshooting information:
!!!!! DO NOT EDIT !!!!!
Name: ‪Circuit Construction Kit: AC‬
URL: https://phet-dev.colorado.edu/html/circuit-construction-kit-ac/1.0.0-dev.31/phet/circuit-construction-kit-ac_en_phet.html
Version: 1.0.0-dev.31 2021-09-28 23:21:59 UTC
Features missing: applicationcache, applicationcache, touch
Flags: pixelRatioScaling
User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.61 Safari/537.36
Language: en-US
Window: 1391x690
Pixel Ratio: 2/1
WebGL: WebGL 1.0 (OpenGL ES 2.0 Chromium)
GLSL: WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor: WebKit (WebKit WebGL)
Vertex: attribs: 16 varying: 31 uniform: 1024
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 80)
Max viewport: 16384x16384
OES_texture_float: true
Dependencies JSON: {}

**Comment by samreid** [comment]
Sept 30, 2021

We discussed why is the Java matrix non-square?  It seems we have the same code generating the equations, but for some reason Java has an extra equation.  I should see where that is coming from.

Kathy asks if we can transform the numbers before solving the matrix, then transform back?

JO: If you can store value imprecision along with the value, you can trace that through the solution.

JO: So this is a precision vs performance tradeoff?
SR: It seems so.

1. Better condititon for the matrices via larger resistances?
KP: We can try multiplying all of those by 100x and see if helps.  But that would get some bad data to show in ammeters or voltmeters.

2. Using QR or another stable solver?
3. More decimal places in the solver?

JO: What if we add other constraints to get an overconstrained solve, like QR?
SR: What about adding "going around a loop gives back to 0 voltage", is that another independent equation?
This seems good to investigate, and see why there is an extra equation in Java?

SR: Should we add all the loops?
Maybe it is tractable if that doesn't become too many, or maybe we would need to prune them somehow.

SR: Maybe these factors would make it so we don't need the poor-performance arbitrary precision solver.

AP: What about treating a large battery like it is multiple small batteries.
SR: Splitting a small resistor into 2 smaller resistors.

JO: Look into the least squares approach.

**Comment by jonathanolson** [comment]
https://math.stackexchange.com/questions/2040363/comparing-lu-or-qr-decompositions-for-solving-least-squares had a note about LU/QR for least squares, notably:

> In cases of ill-conditioning, the orthogonalization methods give an added measure of reliability. 𝐐𝐑 with condition estimate is very dependable and, of course, SVD is unsurpassed when it comes to producing a meaningful solution to a nearly singular system.

**Comment by samreid** [comment]
In https://github.com/phetsims/circuit-construction-kit-dc/issues/96 we made this commit: https://github.com/phetsims/circuit-construction-kit-common/commit/083c00d8bb0af76015c249616e7baa0e7ceb2b4c

```js
      // having charge conservation at each node is overconstraining and causes problems for QR.  In each connected
      // circuit element, we must choose exactly one node at which to avoid the current conservation term.
      if ( referenceNodeIds.indexOf( node ) === -1 ) {

        const incomingCurrentTerms = this.getCurrentTerms( node, 'nodeId1', -1 );
        const outgoingCurrentTerms = this.getCurrentTerms( node, 'nodeId0', +1 );
        const currentConservationTerms = incomingCurrentTerms.concat( outgoingCurrentTerms );
        equations.push( new Equation( this.getCurrentSourceTotal( node ), currentConservationTerms ) );
      }
```

But now perhaps https://github.com/phetsims/circuit-construction-kit-dc/issues/96 will be solved by our new graph algorithms, and we can go back to the overconstrained QR.

**Comment by samreid** [comment]
Summarizing the commits above:
* Added the missing current conservation equation
* Use QR decomposition for the solution
* Avoid the arbitrary-precision solver altogether
* Performance improvements when generating equations
* Increase battery resistance and wire resistivity by 100 (tweakable with query parameters)
* Made the dt threshold less strict.

I published https://phet-dev.colorado.edu/html/circuit-construction-kit-ac/1.0.0-dev.33/phet/circuit-construction-kit-ac_en_phet.html with all these changes, and it's at a good point for review and testing.

**Comment by arouinfar** [comment]
@ariel-phet can you spend some time testing this?

**Comment by ariel-phet** [comment]
@samreid there is currently a severe issue with the behavior involving switches, I believe this video will show you a minimum reproducible case. Kicking back to you until this issue is addressed


https://user-images.githubusercontent.com/4051174/136085612-603b79b6-f889-4d5c-b98d-3c066d7a17a1.mp4

**Comment by samreid** [comment]
I tried to reproduce the problem in dev.33, but didn't see the buggy behavior:

![Kapture 2021-10-05 at 13 06 34](https://user-images.githubusercontent.com/679486/136087054-e94808e4-95c4-4a54-bd91-32c55d709058.gif)
Did you see the problem in dev.33?  Does it happen every time?  Is it dependent on the order of construction?

**Comment by samreid** [comment]
I addressed the "current travels through an open switch" issue in #745, and I published https://phet-dev.colorado.edu/html/circuit-construction-kit-ac/1.0.0-dev.34/phet/circuit-construction-kit-ac_en_phet.html.  That seemed straightforward, so I think testing on this issue can be unblocked.

**Comment by ariel-phet** [comment]
@samreid yes it is dev.34, and same result on firefox.

Try playing around with wire lengths (asymmetric wire lengths seem to be the most obvious)

**Comment by samreid** [comment]
I tested dev.34 on my Win/Chrome 94.0.4606.71.  For the capacitor/ac source with varying wire lengths, it is running a smooth 60fps.

Looks equally speedy on my Win/Firefox 92.0.  So I will need help reproducing this performance degradation so I can investigate it further.

**Comment by samreid** [comment]
I was able to reproduce the problem on my Win/Chrome, mostly using the 3rd circuit in https://github.com/phetsims/circuit-construction-kit-common/issues/744#issuecomment-937946647.  Putting an AC source on the top component also makes the problem worse.  I added a commit that gives a new query parameter `minDT` which defaults to 1E-5.  I'm seeing that turning that up to 1E-3 the performance gets a lot better on Win/Chrome.  Instead of thousands of iterations per frame, it runs  more like 15-20 at the most.  The behavior looks qualitatively correct.  However, at 1E-5, we have 2312 tests passing.  At 1E-4, we have only 1784 assertions of 2312 passed.  At 1E-3, we have 221 assertions of 2312 passed.  This is for a strictness of `assert.ok( error < 1E-4 );` in the unit test. 

Relaxing this to `assert.ok( error < 1E-3 );`, 1E-4 still gives 2291 assertions of 2312 passed, 21 failed..

Relaxing this to `assert.ok( error < 1E-2 );` 1E-4 gives 2299 assertions of 2312 passed, 13 failed.  I think we should look into this to see if these tests are too strict.

**Comment by ariel-phet** [comment]
@samreid so far testing looks good, I played with the various cases I found problems with in dev.34 and all looks good now, and no  obvious performance issues. I also check some cases of parallel batteries, simple circuits, series and parallel resistors, short circuits, RC circuits, RLC circuits, and behavior is looking accurate.  I think things are stable enough now to go play with inductor values and such. Going to unassign myself from this issue for the moment, but leave it open.
